### PR TITLE
Openai

### PR DESCRIPTION
**Pull request by sonichi**
## Why are these changes needed?

Support tuning openai inference hyperparameters.

## Checks

- [x] I've used [pre-commit](https://microsoft.github.io/FLAML/docs/Contribute#pre-commit) to lint the changes in this PR, or I've made sure [lint with flake8](https://github.com/microsoft/FLAML/blob/816a82a1155b4de4705b21a615ccdff67c6da379/.github/workflows/python-package.yml#L54-L59) output is two 0s.
- [ ] I've included any doc changes needed for https://microsoft.github.io/FLAML/. See https://microsoft.github.io/FLAML/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

### CODE REVIEW

File: flaml/integrations/oai/completion.py [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+    """Get a unique identifier of a configuration.
+
+    Args:
+        config (dict or list): A configuration.
+
+    Returns:
+        tuple: A unique identifier which can be used as a key for a dict.
+    """
+    if isinstance(config, dict):
+        return tuple(get_key(x) for x in sorted(config.items()))
```
Comment (liususan091219) [inline flaml/integrations/oai/completion.py:28]: sorted by what?

File: flaml/integrations/oai/completion.py [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+
+    default_search_space = {
+        "model": tune.choice(list(price1K.keys())),
+        "temperature_or_top_p": tune.choice(
+            [
+                {"temperature": tune.uniform(0, 1)},
+                {"top_p": tune.uniform(0, 1)},
+            ]
+        ),
+        "max_tokens": tune.lograndint(100, 1000),
```
Comment (liususan091219) [inline flaml/integrations/oai/completion.py:54]: start from 100? log?

File: test/openai/test_tune.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    ]
+    oai.Completion.set_cache(seed)
+    config, analysis = oai.Completion.tune(
+        data=tune_data,
+        metric="expected_success",
+        mode="max",
+        eval_func=success_metrics,
+        log_file_name="logs/humaneval.log",
+        inference_budget=0.02,
+        optimization_budget=5,
```
Comment (liususan091219) [inline test/openai/test_tune.py:92]: Is the optimization budget required?

**Comment by liususan091219** [comment]
integrations and nlp are the same level; but openai is also nlp, should we reorganize the hierarchy?

File: notebook/integrate_openai.ipynb [hunk 4]
```diff
@@ -0,0 +1,913 @@
...
+   ],
+   "source": [
+    "config, analysis = oai.Completion.tune(\n",
+    "    data=tune_data,  # the data for tuning\n",
+    "    metric=\"expected_success\",  # the metric to optimize\n",
+    "    mode=\"max\",  # the optimization mode\n",
+    "    eval_func=success_metrics,  # the evaluation function to return the success metrics\n",
+    "    # log_file_name=\"logs/humaneval.log\",  # the log file name\n",
+    "    inference_budget=0.02,  # the inference budget (dollar)\n",
+    "    optimization_budget=2,  # the optimization bueget (dollar)\n",
```
Comment (liususan091219) [inline notebook/integrate_openai.ipynb:455]: budget

**Review by liususan091219 — CHANGES_REQUESTED** [review]

File: flaml/integrations/oai/completion.py (hunk 1 above)
Comment (sonichi) [reply flaml/integrations/oai/completion.py:28]: I don't understand this question.

File: flaml/integrations/oai/completion.py (hunk 2 above)
Comment (sonichi) [reply flaml/integrations/oai/completion.py:54]: I don't understand this question.

File: test/openai/test_tune.py (hunk 3 above)
Comment (sonichi) [reply test/openai/test_tune.py:92]: yes

File: notebook/integrate_openai.ipynb (hunk 4 above)
Comment (sonichi) [reply notebook/integrate_openai.ipynb]: fixed.

File: flaml/integrations/oai/completion.py [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+            temperature = space.pop("temperature", None)
+            top_p = space.pop("top_p", None)
+            if temperature is not None and top_p is None:
+                space["temperature_or_top_p"] = {"temperature": temperature}
+            elif temperature is None and top_p is not None:
+                space["temperature_or_top_p"] = {"top_p": top_p}
+            elif temperature is not None and top_p is not None:
+                space.pop("temperature_or_top_p")
+                space["temperature"] = temperature
+                space["top_p"] = top_p
```
Comment (liususan091219) [inline flaml/integrations/oai/completion.py:369]: not recommended by openai

Comment (sonichi) [reply flaml/integrations/oai/completion.py:369]: Yes. This is when users specify temperature and top_p together. What's your suggestion?

File: flaml/integrations/oai/completion.py [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+        return response
+
+    @classmethod
+    def _get_max_safe_n(cls, key, max_tokens):
+        # find the max value in max_safe_n_per_max_tokens whose key is equal or larger than max_tokens
+        return max(
+            (
+                value
+                for key, value in cls._max_safe_n_per_max_tokens.get(key, {}).items()
+                if key >= max_tokens
```
Comment (liususan091219) [inline flaml/integrations/oai/completion.py:120]: is this correct? key is both the input argument and also a local variable

File: flaml/integrations/oai/completion.py [hunk 7]
```diff
@@ -0,0 +1,441 @@
...
+    @classmethod
+    def _get_min_unsafe_n(cls, key, max_tokens):
+        # find the min value in min_unsafe_n_per_max_tokens whose key is equal or smaller than max_tokens
+        return min(
+            (
+                value
+                for key, value in cls._min_unsafe_n_per_max_tokens.get(key, {}).items()
+                if key <= max_tokens
+            ),
+            default=None,
```
Comment (liususan091219) [inline flaml/integrations/oai/completion.py:134]: same here

File: flaml/integrations/oai/completion.py [hunk 8]
```diff
@@ -0,0 +1,441 @@
...
+                        n_tokens_list.append(n_tokens)
+                        responses_list.append(responses)
+                n_tokens = np.mean(n_tokens_list[:data_limit])
+                rho = (
+                    (1 - data_limit / data_length) * (1 + 1 / data_limit)
+                    if data_limit << 1 > data_length
+                    else (1 - (data_limit - 1) / data_length)
+                )
+                # Hoeffding-Serfling bound
+                ratio = 0.1 * np.sqrt(rho / data_limit)
```
Comment (liususan091219) [inline flaml/integrations/oai/completion.py:257]: rho and ratio are not consistent with algorithm 1 in the paper. can you check the correctness?

**Comment by sonichi** [comment]
> integrations and nlp are the same level; but openai is also nlp, should we reorganize the hierarchy?

openai is not exclusive to nlp.

File: flaml/integrations/oai/completion.py (hunk 8 above)
Comment (sonichi) [reply flaml/integrations/oai/completion.py:257]: Checked. The code is correct.

File: flaml/integrations/oai/completion.py (hunk 6 above)
Comment (sonichi) [reply flaml/integrations/oai/completion.py]: revised.

File: flaml/integrations/oai/completion.py (hunk 7 above)
Comment (sonichi) [reply flaml/integrations/oai/completion.py:134]: revised.

File: flaml/integrations/oai/completion.py [hunk 9]
```diff
@@ -0,0 +1,441 @@
...
+                return response
+            except (
+                ServiceUnavailableError,
+                APIError,
+                InvalidRequestError,
+                APIConnectionError,
+            ):
+                logger.info(f"retrying in {cls.retry_time} seconds...", exc_info=1)
+                sleep(cls.retry_time)
+            except RateLimitError:
```
Comment (ekzhu) [inline flaml/integrations/oai/completion.py:103]: Use https://pypi.org/project/ratelimit/ library instead for the whole function.

File: flaml/integrations/oai/completion.py [hunk 10]
```diff
@@ -0,0 +1,441 @@
...
+                        result[metric] = 0
+                        result["cost"] = cost
+                        return result
+                    # evaluate the quality of the responses
+                    responses = [r["text"].rstrip() for r in response["choices"]]
+                    n_tokens = (
+                        response["usage"]["completion_tokens"]
+                        if previous_n
+                        else response["usage"]["total_tokens"]
+                    )
```
Comment (liususan091219) [inline flaml/integrations/oai/completion.py:225]: explain the trick

File: flaml/integrations/oai/completion.py [hunk 11]
```diff
@@ -0,0 +1,441 @@
...
+                    ):
+                        # limit the total tuning cost
+                        return {
+                            metric: 0,
+                            "total_cost": cls._total_cost,
+                            "cost": cost,
+                        }
+                    if previous_n:
+                        n_tokens_list[i] += n_tokens
+                        responses_list[i].extend(responses)
```
Comment (liususan091219) [inline flaml/integrations/oai/completion.py:246]: explain the assumption that the concatenated responses is equal to the one-time response

File: flaml/integrations/oai/completion.py [hunk 12]
```diff
@@ -0,0 +1,441 @@
...
+                    else:
+                        n_tokens_list.append(n_tokens)
+                        responses_list.append(responses)
+                n_tokens = np.mean(n_tokens_list[:data_limit])
```
Comment (liususan091219) [inline flaml/integrations/oai/completion.py:250]:
```suggestion
                avg_n_tokens = np.mean(n_tokens_list[:data_limit])
```

File: flaml/integrations/oai/completion.py [hunk 13]
```diff
@@ -0,0 +1,441 @@
...
+                rho = (
+                    (1 - data_limit / data_length) * (1 + 1 / data_limit)
+                    if data_limit << 1 > data_length
+                    else (1 - (data_limit - 1) / data_length)
+                )
+                # Hoeffding-Serfling bound
+                ratio = 0.1 * np.sqrt(rho / data_limit)
+                if (
+                    target_n_tokens
+                    and n_tokens > target_n_tokens * (1 + ratio)
```
Comment (liususan091219) [inline flaml/integrations/oai/completion.py:260]:
```suggestion
                    and avg_n_tokens > target_n_tokens * (1 + ratio)
```

**Review by liususan091219 — CHANGES_REQUESTED** [review]

File: flaml/integrations/oai/completion.py [hunk 14]
```diff
@@ -0,0 +1,441 @@
...
+                        if prune:
+                            # update unsafe n and prune this config
+                            cls._min_unsafe_n_per_max_tokens[
+                                region_key
+                            ] = unsafe_n = cls._min_unsafe_n_per_max_tokens.get(
+                                region_key, {}
+                            )
+                            unsafe_n[max_tokens] = min(
+                                n, unsafe_n.get(max_tokens, np.inf)
+                            )
```
Comment (liususan091219) [inline flaml/integrations/oai/completion.py:215]: define a function named update_unsafe_n?

File: flaml/integrations/oai/completion.py [hunk 15]
```diff
@@ -0,0 +1,441 @@
...
+                        )
+                        unsafe_n[max_tokens] = min(n, unsafe_n.get(max_tokens, np.inf))
+                    result[metric] = 0
+                    result["total_cost"] = cls._total_cost
+                    result["cost"] = cost
+                    return result
+                if (
+                    prune
+                    and target_n_tokens
+                    and n_tokens <= target_n_tokens * (1 - ratio)
```
Comment (liususan091219) [inline flaml/integrations/oai/completion.py:278]: different from paper

File: notebook/integrate_openai.ipynb [hunk 16]
```diff
@@ -0,0 +1,911 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Define Success Metric\n",
+    "\n",
+    "Before we start tuning, we need to define the success metric we want to opotimize.\n",
```
Comment (liususan091219) [inline notebook/integrate_openai.ipynb:216]:
```suggestion
    "Before we start tuning, we need to define the success metric we want to optimize.\n",
```

**Review by liususan091219 — CHANGES_REQUESTED** [review]

File: flaml/integrations/oai/completion.py (hunk 5 above)
Comment (liususan091219) [reply flaml/integrations/oai/completion.py:369]: print a warning?

File: flaml/integrations/oai/completion.py (hunk 15 above)
Comment (sonichi) [reply flaml/integrations/oai/completion.py:278]: what do you mean?

File: flaml/integrations/oai/completion.py (hunk 9 above)
Comment (sonichi) [reply flaml/integrations/oai/completion.py:103]: Thanks for the suggestion. Let's discuss how to improve it in the next PR. The exceptions handled here are complicated. I don't see a trivial better solution yet.

File: flaml/integrations/oai/completion.py [hunk 17]
```diff
@@ -0,0 +1,453 @@
...
+                        # limit the total tuning cost
+                        return {
+                            metric: 0,
+                            "total_cost": cls._total_cost,
+                            "cost": cost,
+                        }
+                    if previous_num_completions:
+                        n_tokens_list[i] += n_tokens
+                        responses_list[i].extend(responses)
+                        # assuming requesting n1, n2 responses separatively then combining them
```
Comment (liususan091219) [inline flaml/integrations/oai/completion.py:253]:
```suggestion
                        # Assumption 1: assuming requesting n1, n2 responses separatively then combining them
```

File: flaml/integrations/oai/completion.py [hunk 18]
```diff
@@ -0,0 +1,453 @@
...
+                            prune, region_key, max_tokens, num_completions
+                        )
+                        result[metric] = 0
+                        result["cost"] = cost
+                        return result
+                    # evaluate the quality of the responses
+                    responses = [r["text"].rstrip() for r in response["choices"]]
+                    n_tokens = (
+                        response["usage"]["completion_tokens"]
+                        if previous_num_completions  # if querying incrementally, input tokens are not counted
```
Comment (liususan091219) [inline flaml/integrations/oai/completion.py:229]:
```suggestion
                        if previous_num_completions  # Under Assumption 1, we should count both the input and output tokens in the first query, and only count ouput tokens afterwards
```

**Review by liususan091219 — CHANGES_REQUESTED** [review]

**Review by liususan091219 — APPROVED** [review]

File: flaml/integrations/oai/__init__.py [hunk 19]
```diff
@@ -0,0 +1,3 @@
+from .completion import Completion
```
Comment (thinkall) [inline flaml/integrations/oai/__init__.py:1]: `from flaml.integrations.oai.completion import Completion`?

File: notebook/integrate_openai.ipynb [hunk 20]
```diff
@@ -0,0 +1,919 @@
...
+    "\n",
+    "# Use FLAML to Tune OpenAI Models\n",
+    "\n",
+    "In this notebook, we tune OpenAI models for code generation. We use [the HumanEval benchmark](https://huggingface.co/datasets/openai_humaneval) released by OpenAI for synthesizing programs from docstrings. \n",
+    "\n",
+    "## Requirements\n",
+    "\n",
+    "FLAML requires `Python>=3.7`. To run this notebook example, please install flaml with the [openai] option:\n",
+    "```bash\n",
+    "pip install flaml[openai]\n",
```
Comment (thinkall) [inline notebook/integrate_openai.ipynb:24]: Will it work with Ray or Spark parallelization?

Comment (sonichi) [reply notebook/integrate_openai.ipynb:24]: Not added now. Should add it in future. Added "TODO".

File: flaml/integrations/oai/completion.py [hunk 21]
```diff
@@ -0,0 +1,478 @@
...
+
+    default_search_space = {
+        "model": tune.choice(list(price1K.keys())),
+        "temperature_or_top_p": tune.choice(
+            [
+                {"temperature": tune.uniform(0, 1)},
+                {"top_p": tune.uniform(0, 1)},
+            ]
+        ),
+        "max_tokens": tune.lograndint(100, 1000),
```
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/integrations/oai/completion.py]: changed.

File: flaml/integrations/oai/__init__.py (hunk 19 above)
Comment (sonichi) [reply flaml/integrations/oai/__init__.py]: changed.

File: flaml/__init__.py [hunk 22]
```diff
@@ -1,8 +1,13 @@
...
 from flaml.automl import AutoML, logger_formatter
 from flaml.tune.searcher import CFO, BlendSearch, FLOW2, BlendSearchTuner, RandomSearch
 from flaml.onlineml.autovw import AutoVW
 from flaml.version import __version__
-import logging
+
+try:
+    from flaml.integrations import oai
+except ImportError:
+    oai = "please install flaml[openai] option to use the flaml.oai subpackage"
```
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/__init__.py]: addressed.

File: flaml/integrations/oai/completion.py [hunk 23]
```diff
@@ -0,0 +1,478 @@
+from time import sleep
+import logging
+import openai
```
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/integrations/oai/completion.py]: addressed.

**Review by qingyun-wu — APPROVED** [review]

**Comment by qingyun-wu** [comment]
Add documentation/example for this integration in the documentation site (not necessarily in this PR)?

**Review by thinkall — APPROVED** [review]
LGTM. Thanks.

**Pull request merged**